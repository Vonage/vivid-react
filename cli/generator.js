const { cleanupDir, kebab2Camel, capitalize, toCommaSeparatedList, toJsonObjectsList, event2EventDescriptor } = require('./utils')
const { getTemplate, TemplateToken } = require('./templates/templates')
const { writeFileSync } = require('fs')
const { join } = require('path')
const { OutputLanguage } = require('./consts')
const { getPropTypes, getDefaultProps, getProps } = require('./prop.types')

const renderComponent = tag => language => componentName => {
    const flatEventsList = (tag.events || []).map(x => (typeof x === 'string' ? x : x.name ))
    const result = getTemplate('react-component', language)
        .replace(TemplateToken.EVENTS, toJsonObjectsList(flatEventsList.map(event2EventDescriptor)))
        .replace(TemplateToken.PROPERTIES, toCommaSeparatedList(tag.properties))
        .replace(TemplateToken.ATTRIBUTES, toCommaSeparatedList(tag.attributes))
        .replace(TemplateToken.PROP_TYPES, getPropTypes(tag).join(',\n'))
        .replace(TemplateToken.PROPS, getProps(tag).join(',\n'))        
        .replace(TemplateToken.DEFAULT_PROPS, getDefaultProps(tag).join(',\n'))        
        .replace(TemplateToken.TAG_DESCRIPTOR_JSON, JSON.stringify(tag, null, ' '))
        .replace(new RegExp(TemplateToken.COMPONENT_CLASS_NAME, 'g'), componentName)
        .replace(new RegExp(TemplateToken.TAG, 'g'), tag.name)

    return result
}

const generateWrappers = (outputDir, language = OutputLanguage.JavaScript) => (tags) => {
    cleanupDir(outputDir)
    const imports = []
    const exports = []

    for (const tag of tags) {
        const camelizedName = kebab2Camel(tag.name)
        const componentName = capitalize(camelizedName)
        const componentFileExt = `.${language === OutputLanguage.TypeScript ? 'tsx' : language}`
        const componentFileName = `${camelizedName}.g`
        const componentOutputFileName = join(process.cwd(), outputDir, `${componentFileName}${componentFileExt}`)
        imports.push(`import { ${componentName} } from './${componentFileName}'`)
        exports.push(`  ${componentName}`)
        console.info(`Processing ${componentName}...`)

        writeFileSync(
            componentOutputFileName,
            renderComponent(tag)(language)(componentName)
        )
    }

    const indexOutputFileName = join(process.cwd(), outputDir, `index.${language}`)
    writeFileSync(
        indexOutputFileName,
        getTemplate('index', language)
            .replace(TemplateToken.EXPORTS, exports.join(',\n'))
            .replace(TemplateToken.IMPORTS, imports.join('\n'))
    )
    console.info(`${imports.length} wrappers generated at ${outputDir}`)
}

module.exports = {
    generateWrappers
}
