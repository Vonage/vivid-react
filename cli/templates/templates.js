/*
* Purpose: Support code-generation activities with templates
*/
const { readFileSync } = require('fs')

const TemplateToken = {
    BANNER: '<% banner %>',
    IMPORTS: '<% imports %>',
    EXPORTS: '<% exports %>',
    TAG: '<% tag %>',
    COMPONENT_CLASS_NAME: '<% component-class-name %>',
    PROP_TYPES: '<% prop-types %>',
    PROPS: '<% props %>',
    DEFAULT_PROPS: '<% default-props %>',    
    PROPERTIES: '<% properties %>',
    ATTRIBUTES: '<% attributes %>',
    EVENTS: '<% events %>',
    TAG_DESCRIPTOR_JSON: '<% descriptor-json %>'
}

const GENERATED_CODE_BANNER = `/*
* DO NOT EDIT THIS FILE. IT IS GENERATED
*/`
const templatesFolder = `${__dirname}`
const getTemplate = (template, language) =>
    readFileSync(`${templatesFolder}/${language}/${template}.tmpl`, { encoding: 'utf8' })
        .replace(TemplateToken.BANNER, GENERATED_CODE_BANNER)

module.exports = {
    getTemplate,
    TemplateToken,
    GENERATED_CODE_BANNER
}