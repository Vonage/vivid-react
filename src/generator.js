const { EOL } = require('os')
const packageJson = require('../package.json')

const { outputFile, outputJson } = require('fs-extra')

const {
  compileTypescript,
  prepareDir,
  kebab2Camel,
  capitalize,
  toCommaSeparatedList,
  toJsonObjectsList,
  event2EventDescriptor,
  filePath,
  renderJsDoc,
  getIndexFileName,
  getVividPackageName
} = require('./utils')
const { getTemplate, TemplateToken } = require('./templates/templates')
const { join } = require('path')
const { OutputLanguage, FileName } = require('./consts')
const { getPropTypes, getDefaultProps, getProps } = require('./prop.types')

const generateTypings = outputDir => async tags => {
  const distTs = join(FileName.tempFolder, FileName.tempTsFolder)
  await generateWrappers(distTs, OutputLanguage.TypeScript, false, false)(tags)
  await compileTypescript(distTs)(outputDir)
}

const renderComponent = tag => language => componentName => {
  const flatEventsList = (tag.events || []).map(x => (typeof x === 'string' ? x : x.name))
  const result = getTemplate('react-component', language)
    .replace(TemplateToken.CLASS_JSDOC, renderJsDoc(tag))
    .replace(TemplateToken.IMPORTS, `import '${getVividPackageName(tag.path)}'`)
    .replace(TemplateToken.EVENTS, toJsonObjectsList(flatEventsList.map(event2EventDescriptor)))
    .replace(TemplateToken.PROPERTIES, toCommaSeparatedList(tag.properties))
    .replace(TemplateToken.ATTRIBUTES, toCommaSeparatedList(tag.attributes))
    .replace(TemplateToken.PROP_TYPES, getPropTypes(tag).join(',\n'))
    .replace(TemplateToken.PROPS, getProps(tag).join(',\n'))
    .replace(TemplateToken.DEFAULT_PROPS, getDefaultProps(tag).join(',\n'))
    .replace(TemplateToken.TAG_DESCRIPTOR_JSON, JSON.stringify(tag, null, ' '))
    .replace(new RegExp(TemplateToken.COMPONENT_CLASS_NAME, 'g'), componentName)
    .replace(new RegExp(TemplateToken.TAG, 'g'), tag.name)

  return result
}

const getExportLine = componentName => `export { default as ${componentName} } from './${componentName}'`

const generateWrappers = (outputDir, language = OutputLanguage.JavaScript, cleanTemp = true, verbose = true) => async (tags) => {
  const indexFileName = getIndexFileName(language)
  const saveIndex = (outputDir, content) => {
    const indexOutputFileName = join(outputDir, indexFileName)
    return outputFile(
      indexOutputFileName,
      content
    )
  }
  const saveStory = (outputDir, componentName, content) => {
    const indexOutputFileName = join(outputDir, `${componentName}.stories.jsx`)
    return outputFile(
      indexOutputFileName,
      content
    )
  }
  const getIndexContent = (components) =>
    getTemplate('index', language).replace(TemplateToken.EXPORTS, components.map(getExportLine).join(EOL))
  const getStoriesContent = (componentName, tag) =>
    getTemplate('stories', 'js')
      .split(TemplateToken.COMPONENT_CLASS_NAME).join(componentName)

  prepareDir(outputDir, true, verbose)
  const components = []

  for (const tag of tags) {
    const camelizedName = kebab2Camel(tag.name)
    const componentName = capitalize(camelizedName)
    components.push(componentName)
    if (verbose) {
      console.info(`Processing ${componentName}...`)
    }

    const componentOutputDir = join(process.cwd(), outputDir, componentName)
    const storyOutputDir = join(process.cwd(), FileName.storyOutputDir, componentName)
    const componentContent = renderComponent(tag)(language)(componentName)

    await saveIndex(componentOutputDir, componentContent)
    await saveStory(storyOutputDir, componentName, getStoriesContent(componentName, tag))
    console.log('tag.properties', tag.properties)
    console.log('tag', tag.properties)

    const packageName = getVividPackageName(tag.path)
    const packageJsonContent = {
      name: `@vonage/vivid-react-${tag.name}`,
      version: packageJson.version,
      main: indexFileName,
      types: 'index.d.ts',
      private: true,
      license: 'MIT',
      dependencies: {
        [packageName]: packageJson.dependencies[packageName]
      }
    }
    await outputJson(join(componentOutputDir, FileName.packageJson), packageJsonContent, { spaces: 2 })
  }

  await saveIndex(outputDir, getIndexContent(components))

  if (language === OutputLanguage.JavaScript) {
    await generateTypings(outputDir)(tags)
  }

  prepareDir(filePath(FileName.tempFolder), cleanTemp, verbose)

  if (verbose) {
    console.info(`${components.length} wrappers generated at ${outputDir}`)
  }
}

module.exports = {
  generateWrappers
}
