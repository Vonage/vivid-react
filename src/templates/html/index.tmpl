<html>
    <head>
        <script src="https://cdn.vivid.vonage.com/vivid/<% vivid.version %>/all.min.js"></script>
    </head>
    <body>
<% tag %>
    </body>
</html>