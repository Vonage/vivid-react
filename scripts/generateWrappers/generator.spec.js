import fs from "fs-extra";
import { readFileSync } from "fs";
import path from "path";
import { TemplateToken } from "./templates/templates";
import { generateWrappersV3 } from "./generator";

const MOCK_DATA = {
  schemaVersion: "1.0.0",
  readme: "",
  modules: [
    {
      kind: "javascript-module",
      path: "path/to/module",
      declarations: [
        {
          kind: "variable",
          name: "variable1",
          type: {
            text: "{\n\t/**\n\t * Description 1.\n\t */\n\tvalue1: 'value1',\n\n\t/**\n\t * Description 2.\n\t */\n\tvalue2: 'value2',\n}",
          },
          default:
            "{\n\t/**\n\t * Description 1.\n\t */\n\tvalue1: 'value1',\n\n\t/**\n\t * Description 2.\n\t */\n\tvalue2: 'value2',\n}",
          description: "variable1 description",
          privacy: "public",
        },
        {
          kind: "class",
          description: "",
          name: "Class1",
          slots: [
            {
              description: "Default slot.",
              name: "",
            },
          ],
          members: [
            {
              kind: "field",
              name: "field1",
              type: {
                text: "AccordionExpandMode",
              },
              description: "Field1Description",
              privacy: "public",
            },
            {
              kind: "field",
              name: "field2",
              type: {
                text: "string | null",
              },
            },
            {
              kind: "field",
              name: "activeItemIndex",
              type: {
                text: "number",
              },
              default: "0",
            },
            {
              kind: "method",
              name: "method1",
              return: {
                type: {
                  text: "Array | null",
                },
              },
            },
            {
              kind: "method",
              name: "method2",
              return: {
                type: {
                  text: "void",
                },
              },
              parameters: [
                {
                  name: "item",
                  type: {
                    text: "number",
                  },
                },
              ],
            },
          ],
          events: [
            {
              type: {
                text: "CustomEvent<string | null>",
              },
              description:
                "Fires a custom 'change' event when the active item changes",
              name: "change",
            },
          ],
          attributes: [
            {
              name: "field-one",
              type: {
                text: "AccordionExpandMode",
              },
              description: "Attribute1Description",
              fieldName: "field1",
            },
          ],
          superclass: {
            name: "VividElement",
            module:
              "/libs/components/src/shared/foundation/vivid-element/vivid-element",
          },
          vividComponent: {
            public: true,
            name: "accordion",
          },
        },
      ],
      exports: [
        {
          kind: "js",
          name: "AccordionExpandMode",
          declaration: {
            name: "AccordionExpandMode",
            module: "libs/components/src/lib/accordion/accordion.ts",
          },
        },
        {
          kind: "js",
          name: "Class1",
          declaration: {
            name: "Class1",
            module: "libs/components/src/lib/accordion/accordion.ts",
          },
        },
      ],
    },
  ],
};

jest.mock("mkdirp", () => {
  return {
    sync: jest.fn(),
  };
});

jest.mock("fs-extra", () => ({
  outputFile: jest.fn(),
  outputJson: jest.fn(),
}));

jest.mock("fs", () => ({
  readFileSync: jest.fn(),
  existsSync: jest.fn().mockReturnValue(false),
}));

jest.mock("./templates/templates", () => {
  const actualTemplates = jest.requireActual("./templates/templates");
  return {
    ...actualTemplates,
    getTemplate: jest.fn().mockReturnValue(""),
  };
});

describe("generator", () => {
  beforeEach(() => {
    fs.outputFile.mockReset();
    jest.spyOn(console, "info").mockImplementation(jest.fn());
  });

  afterEach(() => {});

  describe("generateWrappersV3", () => {
    function setTokenTemplateMock(templateToken) {
      jest
        .spyOn(require("./templates/templates"), "getTemplate")
        .mockReturnValueOnce(templateToken);
    }
    const generateWrappersV3InstanceJS = generateWrappersV3();
    const generateWrappersV3InstanceTS = generateWrappersV3("", "ts");

    it("should throw if meta is not supplied", async () => {
      await expect(generateWrappersV3InstanceJS()).rejects.toThrow();
      await expect(generateWrappersV3InstanceTS()).rejects.toThrow();
    });

    it("should output a file with the className inside the v3 directory", async () => {
      readFileSync.mockReturnValue(JSON.stringify(MOCK_DATA));
      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });
      expect(fs.outputFile).toHaveBeenCalledWith(
        path.resolve(process.cwd(), "v3/VwcClass1/index.tsx"),
        expect.any(String)
      );
    });

    it("should replace TemplateToken.CLASS_JSDOC with JSDOC content", async () => {
      setTokenTemplateMock(TemplateToken.CLASS_JSDOC);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchSnapshot();
    });

    it("should replace TemplateToken.ATTRIBUTES with empty string", async () => {
      setTokenTemplateMock(TemplateToken.ATTRIBUTES);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("");
    });

    it("should replace TemplateToken.PROP_TYPES with empty string", async () => {
      setTokenTemplateMock(TemplateToken.PROP_TYPES);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("");
    });

    it("should replace TemplateToken.PROPS with empty string", async () => {
      setTokenTemplateMock(TemplateToken.PROPS);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchInlineSnapshot(`
        "  onChange?: (event: SyntheticEvent) => void,
          \\"field1\\"?: any /* AccordionExpandMode */,
          \\"field2\\"?: string | null,
          \\"activeItemIndex\\"?: number"
      `);
    });

    it("should replace TemplateToken.TAG_DESCRIPTOR_JSON with empty string", async () => {
      setTokenTemplateMock(TemplateToken.TAG_DESCRIPTOR_JSON);
      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchSnapshot();
    });

    it("should replace TemplateToken.COMPONENT_CLASS_NAME with empty string", async () => {
      setTokenTemplateMock(TemplateToken.COMPONENT_CLASS_NAME);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("VwcClass1");
    });

    it("should replace TemplateToken.TAG_PREFIX with empty string", async () => {
      setTokenTemplateMock(TemplateToken.TAG_PREFIX);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("vvd3");
    });

    it("should replace TemplateToken.COMPONENT_NAME with empty string", async () => {
      setTokenTemplateMock(TemplateToken.COMPONENT_NAME);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("Class1");
    });

    it("should replace TemplateToken.TAG with empty string", async () => {
      setTokenTemplateMock(TemplateToken.TAG);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("vvd3-class1");
    });

    it("should replace TemplateToken.EVENTS with empty string", async () => {
      setTokenTemplateMock(TemplateToken.EVENTS);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(JSON.parse(content)).toEqual({
        name: "change",
        propName: "onChange",
      });
    });

    it("should replace TemplateToken.PROPERTIES with comma separated list of fields", async () => {
      setTokenTemplateMock(TemplateToken.PROPERTIES);

      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toEqual("'field1', 'field2', 'activeItemIndex'");
    });

    it("should replace TemplateToken.PROPS and remove duplicate values", async () => {
      function pushAnotherFieldWithExistingFieldName() {
        const newData = JSON.parse(JSON.stringify(MOCK_DATA));
        newData.modules[0].declarations[1].members.push({
          kind: "field",
          name: "field1",
          type: {
            text: "AccordionExpandMode",
          },
          description: "Field1Description",
          privacy: "public",
        });
        return newData;
      }

      setTokenTemplateMock(TemplateToken.PROPS);
      const newData = pushAnotherFieldWithExistingFieldName();

      await generateWrappersV3InstanceTS({ elements: newData });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchInlineSnapshot(`
        "  onChange?: (event: SyntheticEvent) => void,
          \\"field1\\"?: any /* AccordionExpandMode */,
          \\"field2\\"?: string | null,
          \\"activeItemIndex\\"?: number"
      `);
    });

    it("should add extra props from Vivid3ComponentsExtraPropertiesMap", async () => {
      function replaceComponentWithOverridable() {
        const newData = JSON.parse(JSON.stringify(MOCK_DATA));
        newData.modules[0].declarations[1].name = 'DataGridCell'
        return newData;
      }
      setTokenTemplateMock(TemplateToken.PROPS);
      const newData = replaceComponentWithOverridable();

      await generateWrappersV3InstanceTS({ elements: newData });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchInlineSnapshot(`
        "  onChange?: (event: SyntheticEvent) => void,
          \\"ariaSelected\\"?: any /* boolean | undefined */,
          \\"field1\\"?: any /* AccordionExpandMode */,
          \\"field2\\"?: string | null,
          \\"activeItemIndex\\"?: number"
      `);
    });
 
    it("should override default props with extra props from Vivid3ComponentsExtraPropertiesMap", async () => {
      function replaceComponentWithOverridable() {
        const newData = JSON.parse(JSON.stringify(MOCK_DATA));
        newData.modules[0].declarations[1].name = 'DataGridCell'
        newData.modules[0].declarations[1].members[0].name = 'ariaSelected'
        newData.modules[0].declarations[1].attributes[0].name = 'ariaSelected'
        return newData;
      }
      setTokenTemplateMock(TemplateToken.PROPS);
      const newData = replaceComponentWithOverridable();

      await generateWrappersV3InstanceTS({ elements: newData });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content).toMatchInlineSnapshot(`
        "  onChange?: (event: SyntheticEvent) => void,
          \\"ariaSelected\\"?: any /* boolean | undefined */,
          \\"field2\\"?: string | null,
          \\"activeItemIndex\\"?: number"
      `);
    });

    it("should replace component registration for DataGridRow with empty string", async () => {
      setTokenTemplateMock(`
            import { register<% component-name %> } from '@vonage/vivid'

            register<% component-name %>('<% tag-prefix %>')
            `);

      MOCK_DATA.modules[0].declarations.find((x) => x.kind === "class").name =
        "DataGridRow";
      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content.trim()).toBe("");
    });

    it("should replace component registration for DataGridCell with empty string", async () => {
      setTokenTemplateMock(`
            import { register<% component-name %> } from '@vonage/vivid'

            register<% component-name %>('<% tag-prefix %>')
            `);

      MOCK_DATA.modules[0].declarations.find((x) => x.kind === "class").name =
        "DataGridCell";
      await generateWrappersV3InstanceTS({ elements: MOCK_DATA });

      const content = fs.outputFile.mock.calls[0][1];
      expect(content.trim()).toBe("");
    });
  });
});
