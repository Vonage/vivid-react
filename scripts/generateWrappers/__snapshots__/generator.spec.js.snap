// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator generateWrappersV3 should replace TemplateToken.CLASS_JSDOC with JSDOC content 1`] = `
"/** VwcClass1 
* For more info on this Vivid element please visit https://vivid.deno.dev/components/class1 
* @param {AccordionExpandMode} field1 - Field1Description **attribute** \`field-one\`
* @param {string | null} field2  
* @param {number} activeItemIndex  
*/"
`;

exports[`generator generateWrappersV3 should replace TemplateToken.TAG_DESCRIPTOR_JSON with empty string 1`] = `
"{
 \\"kind\\": \\"class\\",
 \\"description\\": \\"\\",
 \\"name\\": \\"Class1\\",
 \\"slots\\": [
  {
   \\"description\\": \\"Default slot.\\",
   \\"name\\": \\"\\"
  }
 ],
 \\"members\\": [
  {
   \\"kind\\": \\"field\\",
   \\"name\\": \\"field1\\",
   \\"type\\": {
    \\"text\\": \\"AccordionExpandMode\\"
   },
   \\"description\\": \\"Field1Description\\",
   \\"privacy\\": \\"public\\"
  },
  {
   \\"kind\\": \\"field\\",
   \\"name\\": \\"field2\\",
   \\"type\\": {
    \\"text\\": \\"string | null\\"
   }
  },
  {
   \\"kind\\": \\"field\\",
   \\"name\\": \\"activeItemIndex\\",
   \\"type\\": {
    \\"text\\": \\"number\\"
   },
   \\"default\\": \\"0\\"
  },
  {
   \\"kind\\": \\"method\\",
   \\"name\\": \\"method1\\",
   \\"return\\": {
    \\"type\\": {
     \\"text\\": \\"Array | null\\"
    }
   }
  },
  {
   \\"kind\\": \\"method\\",
   \\"name\\": \\"method2\\",
   \\"return\\": {
    \\"type\\": {
     \\"text\\": \\"void\\"
    }
   },
   \\"parameters\\": [
    {
     \\"name\\": \\"item\\",
     \\"type\\": {
      \\"text\\": \\"number\\"
     }
    }
   ]
  }
 ],
 \\"events\\": [
  {
   \\"type\\": {
    \\"text\\": \\"CustomEvent<string | null>\\"
   },
   \\"description\\": \\"Fires a custom 'change' event when the active item changes\\",
   \\"name\\": \\"change\\"
  }
 ],
 \\"attributes\\": [
  {
   \\"name\\": \\"field-one\\",
   \\"type\\": {
    \\"text\\": \\"AccordionExpandMode\\"
   },
   \\"description\\": \\"Attribute1Description\\",
   \\"fieldName\\": \\"field1\\"
  }
 ],
 \\"superclass\\": {
  \\"name\\": \\"VividElement\\",
  \\"module\\": \\"/libs/components/src/shared/foundation/vivid-element/vivid-element\\"
 },
 \\"vividComponent\\": {
  \\"public\\": true,
  \\"name\\": \\"accordion\\"
 }
}"
`;
